src/gerbil/runtime/c3.ss

```diff
@@ -1,84 +1,181 @@
-;;; for its ordering consistency properties:
-;;; OpenDylan, Python, Raku, Parrot, Solidity, PGF/TikZ.
+;;; We add support for classes that demand single-inheritance:
+;;;  4. The precedence-list of a struct is the tail of the precedence-list of
+;;;     its subclasses.
+;;;
+;;; Additionally, C3 uses the following heuristic to deterministically compute
+;;; the list, starting from the front (most specific), to chose the next
+;;; element of the precedence list when the above rules leave some ambiguity:
+;;; among the possible candidates to be chosen next, pick the one that appears
+;;; earliest in the list of precedence lists followed by the super list.
+;;;
+;;; C3 has since been adopted for multiple inheritance by many modern
+;;; object systems for its ordering consistency properties:
+;;; OpenDylan, Python 2.3, Raku, Parrot, Scala, Solidity, PGF/TikZ.
 ;;;       https://en.wikipedia.org/wiki/C3_linearization
 ;;;       https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.19.3910
 ;;;
+;;; We dub C4 our extension with the 4th constraints for single-inheritance.
+;;;
 ;;; See tests in ../test/c3-test.ss
 
 prelude: "../prelude/core"
 package: gerbil/runtime
 namespace: #f
 
-(export c3-linearize)
+(export c4-linearize c3-linearize)
 (import "util")
 
-;; C3 linearization algorithm: given a top object x from which to compute the precedence list,
+;; C4 linearization algorithm: given a top object x from which to compute the precedence list,
 ;; - rhead is [x] or [] depending on whether to include x as head of the result.
 ;; - supers is (get-supers x), the list of direct supers of x
 ;; - get-precedence-list gets the precedence list for a super s, including s itself in front
+;; - single-inheritance? is a predicate that tells if a class follows single inheritance, and
+;;   must have its precedence list be a tail of any subclass'es precedence list.
 ;; - eqpred is an equality predicate between list elements
 ;;   (passed around because somehow equal? doesn't work correctly on runtime type-descriptors).
 ;; - get-name gets the name of a object/class, for debugging only.
-;; : (List X) (List X) (X -> (NonEmptyList X)) ?(X X -> Bool) ?(X -> Y) -> (List X)
-(def (c3-linearize rhead supers get-precedence-list (eqpred eq?) (get-name identity))
-  (let (tails (map get-precedence-list supers)) ;; : (List (NonEmptyList X))
-    (append1! tails supers)
-    (c3-linearize-loop rhead tails eqpred get-name)))
+;; Returns the linearized precedence list, and the most specific struct superclass if any
+;; : (List X) (List X) \
+;;  get-precedence-list: (X -> (NonEmptyList X)) \
+;;  single-inheritance?: (X -> Bool) \
+;;  eqpred: ?(X X -> Bool) \
+;;  get-name: ?(X -> Y) \
+;; -> (List X) (OrFalse X)
+(def (c4-linearize rhead supers
+                   get-precedence-list: get-precedence-list
+                   single-inheritance?: single-inheritance?
+                   eqpred: (eqpred eq?)
```

Comment:
this is called `test:` in other places (eg hash tables), maybe we should call it test here too.